### PR TITLE
fix(cloud-save): issue with %GameInstall% being prefixed in fileList.files

### PR DESCRIPTION
**Pull request by silentrald**
---
## Summary by cubic
Fixes Steam Auto Cloud path parsing when %GameInstall% is embedded in filenames and when the path is ".". Ensures files are placed correctly during sync and upload.

- **Bug Fixes**
  - Map filenames starting with %GameInstall% to the GameInstall directory and strip the placeholder to build the full path.
  - Treat "." as root: omit "." when building prefix and handle prefixes ending with "%" when creating prefixPath.
  - Add regression tests for both download and upload flows; verify file placement and change number updates.

<sup>Written for commit 3bb48d508863d041975ee1e825faa4048945f1ef. Summary will update on new commits.</sup>

## Summary by CodeRabbit

* **Bug Fixes**
  * Fix path resolution for filenames that include game-install placeholders so files resolve to the correct install location.
  * Improve handling of relative/current-path prefixes to avoid incorrect path construction and misplaced files during sync.
  * Ensure downloads and cloud sync consistently place files and update local change tracking.

* **Tests**
  * Add regression test covering the game-install filename edge case.

### CODE REVIEW

File: app/src/main/java/app/gamenative/service/SteamAutoCloud.kt [hunk 1]
```diff
@@ -143,7 +143,15 @@ object SteamAutoCloud {
...
 
-        val getFullFilePath: (AppFileInfo, AppFileChangeList) -> Path = { file, fileList ->
+        val getFullFilePath: (AppFileInfo, AppFileChangeList) -> Path = getFullFilePath@{ file, fileList ->
+            if (file.filename.startsWith("%${PathType.GameInstall.name}%")) {
+                // Edge case if the files are not properly parsed by the Steam API Call
+                return@getFullFilePath Paths.get(
+                    prefixToPath(PathType.GameInstall.name),
+                    file.filename.substring(PathType.GameInstall.name.length + 2)
+                )
+            }
```
[the remark this answers is not in the document]
Comment (utkarshdalal) [reply app/src/main/java/app/gamenative/service/SteamAutoCloud.kt:153]: @silentrald this is correct, the path is going to start with `/` - That doesn't look intentional.

Why not simply do a `file.filename.replace("%${PathType.GameInstall.name}%\", "")`?

File: app/src/main/java/app/gamenative/data/UserFileInfo.kt [hunk 2]
```diff
@@ -21,12 +21,15 @@ data class UserFileInfo(
...
     val prefix: String
-        get() = Paths.get("%${root.name}%$path").pathString
+        get() = Paths.get("%${root.name}%${ if (path == ".") "" else path }").pathString
             .replace("{64BitSteamID}", SteamUtils.getSteamId64().toString())
             .replace("{Steam3AccountID}", SteamUtils.getSteam3AccountId().toString())
 
     val prefixPath: String
-        get() = Paths.get(prefix, filename).pathString
+        get() = (
+            if (prefix.endsWith('%')) "$prefix$filename"
```
Comment (utkarshdalal) [inline app/src/main/java/app/gamenative/data/UserFileInfo.kt:30]: why are we checking if the prefix ends with %?

File: app/src/test/java/app/gamenative/service/SteamAutoCloudTest.kt [hunk 3]
```diff
@@ -849,5 +852,269 @@ class SteamAutoCloudTest {
...
+                root = PathType.GameInstall,
+                path = ".",
+                pattern = "save1.sav",
+            ),
+            SaveFilePattern(
+                root = PathType.GameInstall,
+                path = ".",
+                pattern = "save2.sav",
+            ),
+        )
```
Comment (utkarshdalal) [inline app/src/test/java/app/gamenative/service/SteamAutoCloudTest.kt:1046]: i'm not following why path has to be `.`? What are we introducing that's new that's making the path `.`?

File: app/src/main/java/app/gamenative/data/UserFileInfo.kt [hunk 4]
```diff
@@ -21,12 +21,15 @@ data class UserFileInfo(
     val sha: ByteArray,
 ) {
     val prefix: String
-        get() = Paths.get("%${root.name}%$path").pathString
+        get() = Paths.get("%${root.name}%${ if (path == ".") "" else path }").pathString
```
Comment (utkarshdalal) [inline app/src/main/java/app/gamenative/data/UserFileInfo.kt:24]: Why is the path becoming `.` and not an empty string?

File: app/src/main/java/app/gamenative/data/UserFileInfo.kt (hunk 2 above)
Comment (silentrald) [reply app/src/main/java/app/gamenative/data/UserFileInfo.kt:30]: `Path.get` appends the slashes even if the pathing which breaks the pathing. ie.

`%GameInstall%/save0.dat`

That's why I added it there to remove it, unless we edit the prefixToPath call to handle this case.

File: app/src/main/java/app/gamenative/service/SteamAutoCloud.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (silentrald) [reply app/src/main/java/app/gamenative/service/SteamAutoCloud.kt:153]: Yeah, I forgot that you can do this, I'll apply this in another commit

File: app/src/main/java/app/gamenative/data/UserFileInfo.kt (hunk 4 above)
Comment (silentrald) [reply app/src/main/java/app/gamenative/data/UserFileInfo.kt:24]: Seems like in

https://github.com/silentrald/GameNative/blob/3bb48d508863d041975ee1e825faa4048945f1ef/app/src/main/java/app/gamenative/service/SteamAutoCloud.kt#L220C1-L237C51

`userFile.substitutedPath` returned here is the `.`. If this is the case would you prefer the handling of the dot in the `SteamAutoCloud` class instead?

File: app/src/main/java/app/gamenative/data/UserFileInfo.kt [hunk 5]
```diff
@@ -21,12 +21,15 @@ data class UserFileInfo(
...
             .replace("{Steam3AccountID}", SteamUtils.getSteam3AccountId().toString())
 
     val prefixPath: String
-        get() = Paths.get(prefix, filename).pathString
+        get() = (
+            if (prefix.endsWith('%')) "$prefix$filename"
+            else Paths.get(prefix, filename).pathString
+        )
             .replace("{64BitSteamID}", SteamUtils.getSteamId64().toString())
             .replace("{Steam3AccountID}", SteamUtils.getSteam3AccountId().toString())
```
[the remark this answers is not in the document]
Comment (utkarshdalal) [reply app/src/main/java/app/gamenative/data/UserFileInfo.kt:34]: that doesn't sound right. How would prefix.endsWith('%') and path == "." be equivalent in any way?

File: app/src/main/java/app/gamenative/data/UserFileInfo.kt (hunk 2 above)
Comment (utkarshdalal) [reply app/src/main/java/app/gamenative/data/UserFileInfo.kt:30]: I see. This change is a bit scary to me, because when we're uploading it will see that file as deleted and delete it from Steam's remote storage.
For instance, it will consider `%GameInstall%/save0.dat` deleted, and try to upload a new file called `%GameInstall%save0.dat`

Comment (utkarshdalal) [reply app/src/main/java/app/gamenative/data/UserFileInfo.kt:30]: Are you sure this is required?

Comment (utkarshdalal) [reply app/src/main/java/app/gamenative/data/UserFileInfo.kt:30]: For instance, I see file paths in the existing tests that look like:
`"%WinMyDocuments%/My Games/TestGame/Steam/76561198025127569"`

So was the upload failing for you when you had a `/`? Can you please check again?

File: app/src/main/java/app/gamenative/data/UserFileInfo.kt (hunk 4 above)
Comment (utkarshdalal) [reply app/src/main/java/app/gamenative/data/UserFileInfo.kt:24]: Got it. Looks like sometimes Steam sends "." instead of "" as the path.

**Comment by silentrald** [comment]
Applied in #508. Closing PR, thanks @utkarshdalal
